[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/mohsennasab/python-fundamentals-hh/blob/main/notebooks/04_raster-data/Module4_RasterData.ipynb)

# Module 4: Raster Data Analysis for H&H Modeling
## Understanding How Water Moves Across the Landscape

### Welcome to Raster Analysis!
This module introduces **raster (gridded) data analysis** - essential for understanding **how** water moves across terrain and **what** it encounters. We'll work with elevation and land cover data to complete your spatial analysis toolkit.

### What You'll Work Through Today:
- Understand raster data as "smart grids" with values  
- Load and visualize Digital Elevation Models (DEMs)  
- Analyze land cover patterns using NLCD data  
- Clip rasters to watershed boundaries  
- Calculate terrain and land use statistics  
- Create elevation bands and slope maps  
- Answer: "How does terrain affect runoff?" and "What land uses dominate?"  

### Module Structure:
1. **Mental Models** - Raster vs Vector thinking
2. **Raster Basics** - Pixels, bands, and values
3. **Loading Raster Data** - DEMs and Land Cover
4. **Clipping to Watersheds** - Focus your analysis
5. **Elevation Analysis** - Terrain statistics
6. **Land Cover Analysis** - Understanding the surface
7. **Raster Calculations** - Slope, aspect, and more
8. **Professional Visualization** - Maps that communicate
9. **What You Can Now Do** - Complete spatial toolkit

---

## Part 1: Mental Models - Thinking in Grids 🧠

### Raster Data: The Landscape as a Grid

**Think of raster data like a spreadsheet covering the Earth:**
- Each cell (pixel) has a value
- Values represent measurements (elevation, rainfall, temperature)
- Cells are uniformly sized and aligned
- Perfect for continuous phenomena

### Vector vs Raster: Two Ways to See the World

| Aspect | Vector (Module 3) | Raster (Module 4) |
|--------|------------------|-------------------|
| **Structure** | Points, lines, polygons | Grid of cells |
| **Best for** | Discrete features | Continuous surfaces |
| **H&H Examples** | Watersheds, pipes, buildings | Elevation, precipitation, land cover |
| **Question answered** | WHERE are things? | HOW does it vary? |

### Today's Engineering Questions

For HUC-12 watershed 101900090108:
1. **What's the terrain like?** → Elevation statistics
2. **How steep is it?** → Slope calculations
3. **What land covers exist?** → NLCD analysis
4. **How does land use affect runoff?** → Curve number assignment

### Key Concepts

**Resolution**: Size of each pixel (e.g., 10m x 10m)
- Higher resolution = smaller pixels = more detail
- Trade-off: Detail vs file size

**Bands**: Layers of information
- DEM: Single band (elevation)
- Land cover: Single band (class codes)
- Satellite imagery: Multiple bands (RGB, infrared)

**NoData**: Pixels with no value
- Outside study area
- Missing measurements
- Must handle in calculations

### The Power of Combining Vector + Raster

**Module 3 (Vector)** told us WHERE the watershed is  
**Module 4 (Raster)** tells us WHAT'S INSIDE and HOW it behaves

Together = Complete spatial understanding for H&H modeling!

## A Note on Using AI in This Module

Raster analysis introduces quite a few new libraries and concepts (rasterio, numpy arrays, coordinate transforms). If a code block looks intimidating, don't worry - just copy it into your AI assistant and ask:

- *"Can you walk me through what this code does step by step?"*
- *"Why do we need to mask NoData values before calculating statistics?"*
- *"What does `rasterio.mask.mask()` actually do?"*

The raster functions have some unfamiliar syntax, but the concepts map directly to things you already know from working with GIS. When in doubt, ask.

## Part 2: Setting Up Our Raster Workspace 🛠️

### Installing Required Libraries

For raster work, we need specialized tools:
- **Rasterio**: Reading/writing raster data
- **Rioxarray**: Raster operations with xarray
- **Rasterstats**: Zonal statistics
- **GeoPandas**: For vector boundaries

In [1]:
# Install raster analysis libraries
!pip install -q rasterio rioxarray rasterstats geopandas folium

print("✅ Raster libraries installed!")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 kB 2.9 MB/s eta 0:00:00
✅ Raster libraries installed!


If you see any warnings during installation, that's usually fine. As long as you see the "installed!" message, you're good to go.

In [ ]:
# Import all necessary libraries
import numpy as np
import pandas as pd
import geopandas as gpd
import rasterio
from rasterio.mask import mask
from rasterio.plot import show
import rioxarray as rxr
from rasterstats import zonal_stats
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
from matplotlib.colors import ListedColormap
import folium
import warnings
warnings.filterwarnings('ignore')

# Set default plot size and keep titles in a regular (non-bold) weight
plt.rcParams['figure.figsize'] = (12, 8)
plt.rcParams['axes.titleweight'] = 'normal'
plt.rcParams['figure.titleweight'] = 'normal'

print("Libraries imported successfully!")
print(f"Rasterio version: {rasterio.__version__}")

### What did we just load?

- **rasterio** reads and writes raster files (GeoTIFFs). Think of it as the raster equivalent of GeoPandas.
- **numpy** handles the actual grid data as arrays. Every raster you load becomes a numpy array under the hood.
- **rasterstats** calculates statistics for raster data within vector boundaries (zonal statistics).
- **matplotlib** and its `patches` module help us build maps with custom legends.

If you want to know more about any of these, ask your AI assistant: *"What is rasterio used for and how does it compare to ArcPy for raster analysis?"*

### Understanding Our Data Files

We'll work with three Wyoming datasets:

1. **Digital Elevation Model (DEM)**
   - File: `USGS_13_n42w105_2021061_DEM.tif`
   - Contains: Elevation values in meters
   - Use: Terrain analysis, slope, flow direction

2. **National Land Cover Database (NLCD)**
   - File: `NLCD2024.tif`
   - Contains: Land use/land cover classes
   - Use: Curve numbers, imperviousness

3. **Watershed Boundaries** (from Module 3)
   - File: `NHD__Watershed_Boundaries_HUC_12_Selected.zip`
   - Contains: HUC-12 polygons
   - Use: Clipping boundary

## Part 3: Loading Data into Colab 📁

Upload your raster and vector files:

In [4]:
# Upload files to Colab
from google.colab import files

print("📤 Please upload these files:")
print("1. USGS_13_n42w105_2021061_DEM.tif (Digital Elevation Model)")
print("2. NLCD2024.tif (Land Cover)")
print("3. NHD__Watershed_Boundaries_HUC_12_Selected.zip (Watersheds)")
print("\nClick 'Choose Files' below and select all three files...")

uploaded = files.upload()

print(f"\n✅ Uploaded {len(uploaded)} files:")
for filename in uploaded.keys():
    print(f"   - {filename}")

📤 Please upload these files:
1. USGS_13_n42w105_2021061_DEM.tif (Digital Elevation Model)
2. NLCD2024.tif (Land Cover)
3. NHD__Watershed_Boundaries_HUC_12_Selected.zip (Watersheds)

Click 'Choose Files' below and select all three files...


Saving NHD__Watershed_Boundaries_HUC_12_Selected.zip to NHD__Watershed_Boundaries_HUC_12_Selected (1).zip
Saving NLCD2024.tif to NLCD2024.tif
Saving USGS_13_n42w105_2021061_DEM.tif to USGS_13_n42w105_2021061_DEM.tif

✅ Uploaded 3 files:
   - NHD__Watershed_Boundaries_HUC_12_Selected (1).zip
   - NLCD2024.tif
   - USGS_13_n42w105_2021061_DEM.tif


## Part 4: Loading and Exploring Raster Data 🗺️

### Reading the Digital Elevation Model

In [5]:
# Open and explore the DEM
dem_path = 'USGS_13_n42w105_2021061_DEM.tif'

with rasterio.open(dem_path) as src:
    print("DEM PROPERTIES")
    print("="*40)
    print(f"Dimensions: {src.width} x {src.height} pixels")
    print(f"Number of bands: {src.count}")
    print(f"Coordinate system: {src.crs}")
    print(f"Pixel size: {src.res[0]:.2f} x {src.res[1]:.2f} {src.crs.linear_units}")
    print(f"Bounds: {src.bounds}")
    print(f"Data type: {src.dtypes[0]}")

    # Read the elevation data
    dem_data = src.read(1)  # Read first (and only) band
    dem_transform = src.transform
    dem_crs = src.crs

print(f"\nElevation range: {dem_data.min():.1f} to {dem_data.max():.1f} meters")
print(f"Mean elevation: {dem_data.mean():.1f} meters")

DEM PROPERTIES
Dimensions: 2041 x 1061 pixels
Number of bands: 1
Coordinate system: EPSG:4269
Pixel size: 0.00 x 0.00 unknown
Bounds: BoundingBox(left=-104.91203703717991, bottom=41.0402777771842, right=-104.72305555549514, top=41.13851851803062)
Data type: float32

Elevation range: -999999.0 to 1963.7 meters
Mean elevation: -316333.4 meters


### What are we looking at?

The output above tells you the key properties of this DEM:
- **Dimensions** (width x height) is the grid size in pixels
- **Pixel size** is the ground distance each pixel covers
- **CRS** is the coordinate reference system (this DEM is likely in geographic coordinates, meaning the pixel size is in degrees)
- **Bounds** tell you the geographic extent

The `src.read(1)` call reads the first (and only) band into a numpy array. From here on, `dem_data` is just a grid of numbers where each number is an elevation in meters.

### Quick Visualization of DEM

In [ ]:
# Visualize the DEM
fig, ax = plt.subplots(figsize=(10, 8))

# Mask NoData values before plotting
dem_display = np.ma.masked_where(dem_data == -999999.0, dem_data)

# Show elevation in feet
dem_display_ft = dem_display * 3.281
im = ax.imshow(dem_display_ft, cmap='terrain', aspect='equal')

# Colorbar in feet, placed on the left side of the plot
plt.colorbar(im, ax=ax, label='Elevation (ft)', shrink=0.7, location='left')

ax.set_title('Digital Elevation Model - Full Extent', fontsize=14)

# Pixel row/column ticks are not meaningful here, so hide them
ax.set_xticks([])
ax.set_yticks([])

plt.tight_layout()
plt.show()

### Reading the plot

The colorbar on the left maps colors to elevation, shown here in feet. Lighter, warmer colors are higher elevations; darker, cooler colors are lower. The pixel row and column ticks have been turned off because they don't carry any real-world meaning.

If you want to experiment with different color schemes, try changing `cmap='terrain'` to `cmap='viridis'` or `cmap='coolwarm'` in the code above.

You can ask your AI assistant: *"What are some good matplotlib colormaps for elevation data, and how do I change the one in this plot?"*

### Reading the Land Cover Data

In [11]:
# Open and explore NLCD
nlcd_path = 'NLCD2024.tif'

with rasterio.open(nlcd_path) as src:
    print("NLCD PROPERTIES")
    print("="*40)
    print(f"Dimensions: {src.width} x {src.height} pixels")
    print(f"Coordinate system: {src.crs}")
    print(f"Pixel size: {src.res[0]:.2f} x {src.res[1]:.2f} {src.crs.linear_units}")
    print(f"Bounds: {src.bounds}")

    # Read the land cover data
    nlcd_data = src.read(1)
    nlcd_transform = src.transform
    nlcd_crs = src.crs

# Get unique land cover values
unique_values = np.unique(nlcd_data[nlcd_data > 0])  # Exclude 0 (NoData)
print(f"\nUnique land cover codes: {unique_values}")
print(f"Number of land cover classes: {len(unique_values)}")

NLCD PROPERTIES
Dimensions: 535 x 354 pixels
Coordinate system: PROJCS["AEA_WGS84",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4326"]],PROJECTION["Albers_Conic_Equal_Area"],PARAMETER["latitude_of_center",23],PARAMETER["longitude_of_center",-96],PARAMETER["standard_parallel_1",29.5],PARAMETER["standard_parallel_2",45.5],PARAMETER["false_easting",0],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH]]
Pixel size: 30.00 x 30.00 metre
Bounds: BoundingBox(left=-741915.0, bottom=2037675.0, right=-725865.0, top=2048295.0)

Unique land cover codes: [ 11  21  22  23  24  31  52  71  81  82  90  95 250]
Number of land cover classes: 13


### What are these land cover codes?

NLCD uses integer codes to represent land cover classes. For example, `21` means "Developed, Open Space" and `42` means "Evergreen Forest." The next cell defines a lookup table that maps these codes to human-readable names and colors.

For the full list of codes and a description of each class, see the [NLCD Land Cover Class Legend and Description](https://www.mrlc.gov/data/legends/national-land-cover-database-class-legend-and-description) from MRLC.

The `unique_values` line shows you which land cover types actually exist in your data - you won't always see all 16 possible classes.

### Define NLCD Classification Scheme

In [12]:
# NLCD Classification codes and descriptions
nlcd_classes = {
    11: 'Open Water',
    12: 'Perennial Ice/Snow',
    21: 'Developed, Open Space',
    22: 'Developed, Low Intensity',
    23: 'Developed, Medium Intensity',
    24: 'Developed, High Intensity',
    31: 'Barren Land',
    41: 'Deciduous Forest',
    42: 'Evergreen Forest',
    43: 'Mixed Forest',
    52: 'Shrub/Scrub',
    71: 'Grassland/Herbaceous',
    81: 'Pasture/Hay',
    82: 'Cultivated Crops',
    90: 'Woody Wetlands',
    95: 'Emergent Herbaceous Wetlands'
}

# Colors for visualization (standard NLCD colors)
nlcd_colors = {
    11: '#466B9F',  # Open Water - Blue
    12: '#D1DEF8',  # Ice/Snow - Light blue
    21: '#DEC5C5',  # Developed Open - Light pink
    22: '#D99282',  # Developed Low - Light red
    23: '#EB0000',  # Developed Medium - Red
    24: '#AB0000',  # Developed High - Dark red
    31: '#B3AC9F',  # Barren - Gray
    41: '#68AB5F',  # Deciduous - Light green
    42: '#1C5F2C',  # Evergreen - Dark green
    43: '#B5C58F',  # Mixed Forest - Yellow-green
    52: '#CCB879',  # Shrub - Tan
    71: '#DFDFC2',  # Grassland - Light yellow
    81: '#D1D182',  # Pasture - Yellow
    82: '#A3CC51',  # Crops - Light green
    90: '#82BA9E',  # Woody Wetlands - Teal
    95: '#DCCCC0'   # Herbaceous Wetlands - Light brown
}

print("NLCD land cover classes loaded:")
for code, desc in nlcd_classes.items():
    if code in unique_values:
        print(f"  {code}: {desc}")

NLCD land cover classes loaded:
  11: Open Water
  21: Developed, Open Space
  22: Developed, Low Intensity
  23: Developed, Medium Intensity
  24: Developed, High Intensity
  31: Barren Land
  52: Shrub/Scrub
  71: Grassland/Herbaceous
  81: Pasture/Hay
  82: Cultivated Crops
  90: Woody Wetlands
  95: Emergent Herbaceous Wetlands


### Why two dictionaries?

`nlcd_classes` maps code numbers to names (for labels and tables), while `nlcd_colors` maps them to hex color codes (for consistent map styling). These are the standard NLCD colors you'd see in any GIS software.

If you're working with a different land cover dataset, you'd just update these dictionaries to match your classification scheme.

## Part 5: Loading Watershed Boundaries 🗂️

We'll use the same watershed from Module 3:

In [13]:
# Load watershed boundaries
watersheds = gpd.read_file('zip://NHD__Watershed_Boundaries_HUC_12_Selected.zip')

print(f"Loaded {len(watersheds)} watersheds")
print(f"CRS: {watersheds.crs}")

# Find our target watershed (same as Module 3)
target_huc = '101900090108'

# Find HUC column
huc_column = None
for col in watersheds.columns:
    if 'huc' in col.lower() or 'HUC' in col:
        huc_column = col
        break

if huc_column:
    target_watershed = watersheds[watersheds[huc_column] == target_huc]
    if len(target_watershed) > 0:
        print(f"\n✅ Found target watershed: {target_huc}")
    else:
        print(f"⚠️ Watershed {target_huc} not found. Using first watershed.")
        target_watershed = watersheds.iloc[[0]]
else:
    print("⚠️ No HUC column found. Using first watershed.")
    target_watershed = watersheds.iloc[[0]]

# Display watershed info
print(f"\nTarget watershed properties:")
for col in target_watershed.columns:
    if col != 'geometry':
        print(f"  {col}: {target_watershed[col].values[0]}")

Loaded 19 watersheds
CRS: EPSG:3857

✅ Found target watershed: 101900090108

Target watershed properties:
  OBJECTID_1: 432
  TNMID: {0033630A-499B-4BB8-B14E-8E5764ED7848}
  MetaSource: {FF24B7E9-B918-4635-ADDB-267C27FDD7E0}
  SourceData: None
  SourceOrig: None
  SourceFeat: None
  LoadDate: 2017-07-20T00:00:00.000
  GNIS_ID: 0
  AreaAcres: 12016.58
  AreaSqKm: 48.63
  States: WY
  HUC12: 101900090108
  Name: Town of South Greeley
  HUType: S
  HUMod: NM
  ToHUC: 101900090109
  NonContrib: 0.0
  NonContr_1: 0.0
  Shape_Leng: 46158.0293969
  Shape_Area: 85681954.8029


## Part 6: Aligning Coordinate Systems 🌍

Critical step: Ensure all data uses the same CRS!

In [14]:
# Check and align CRS
print("Current Coordinate Reference Systems:")
print(f"  DEM: {dem_crs}")
print(f"  NLCD: {nlcd_crs}")
print(f"  Watersheds: {watersheds.crs}")

# Reproject watershed to match DEM CRS
if watersheds.crs != dem_crs:
    print(f"\n📐 Reprojecting watershed to match DEM CRS...")
    target_watershed = target_watershed.to_crs(dem_crs)
    print("✅ All data now in same coordinate system!")
else:
    print("\n✅ All data already in same coordinate system!")

Current Coordinate Reference Systems:
  DEM: EPSG:4269
  NLCD: PROJCS["AEA_WGS84",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4326"]],PROJECTION["Albers_Conic_Equal_Area"],PARAMETER["latitude_of_center",23],PARAMETER["longitude_of_center",-96],PARAMETER["standard_parallel_1",29.5],PARAMETER["standard_parallel_2",45.5],PARAMETER["false_easting",0],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH]]
  Watersheds: EPSG:3857

📐 Reprojecting watershed to match DEM CRS...
✅ All data now in same coordinate system!


### Why does CRS alignment matter here?

To clip a raster with a vector boundary, both datasets need to be in the same coordinate system. If the DEM is in EPSG:4326 (degrees) and your watershed is in a UTM projection (meters), the clip operation won't find any overlap and you'll get an empty result.

The code above reprojects the watershed polygon to match the DEM's CRS. We reproject the vector (fast, just recalculates coordinates) rather than the raster (slow, requires resampling every pixel).

## Part 7: Clipping Rasters to Watershed 🔧

### The Engineering Question

**"What are the terrain and land cover characteristics within our watershed?"**

We'll clip both rasters to focus on our area of interest.

### Clip the DEM

In [17]:
# Clip DEM to watershed boundary
from rasterio.mask import mask
import json

# Get watershed geometry
watershed_geom = target_watershed.geometry.iloc[0]

# Clip the DEM
with rasterio.open(dem_path) as src:
    # Perform the clip
    dem_clipped, dem_clip_transform = mask(
        src,
        [watershed_geom], # Pass the shapely geometry directly in a list
        crop=True,
        nodata=-9999
    )

    # Update metadata for clipped raster
    dem_clip_meta = src.meta.copy()
    dem_clip_meta.update({
        'height': dem_clipped.shape[1],
        'width': dem_clipped.shape[2],
        'transform': dem_clip_transform,
        'nodata': -9999
    })

# Extract the data array (first band)
dem_clipped_data = dem_clipped[0]

# Mask NoData values
dem_clipped_masked = np.ma.masked_where(dem_clipped_data == -9999, dem_clipped_data)

print(f"DEM clipped to watershed:")
print(f"  Original size: {dem_data.shape}")
print(f"  Clipped size: {dem_clipped_data.shape}")
print(f"  Valid pixels: {(~dem_clipped_masked.mask).sum():,}")

DEM clipped to watershed:
  Original size: (1061, 2041)
  Clipped size: (703, 1569)
  Valid pixels: 607,809


### What just happened?

`rasterio.mask.mask()` is the key function here. It takes a raster file and a polygon, and returns only the pixels that fall inside that polygon. Everything outside gets set to the NoData value (-9999).

Think of it like using a cookie cutter on a sheet of data - you keep the shape you care about and discard the rest.

The `crop=True` parameter also trims the output array to the bounding box of the polygon, so you're not carrying around a huge array that's mostly NoData.

Want to understand the transform and metadata? Ask your AI: *"What is a rasterio transform and why do we need to update it after clipping?"*

### Clip the Land Cover

In [19]:
# Clip NLCD to watershed boundary
with rasterio.open(nlcd_path) as src:
    # First reproject watershed to NLCD CRS if needed
    watershed_for_nlcd = target_watershed.to_crs(src.crs)
    # Directly get the shapely geometry object for clipping, avoiding JSON serialization issue
    watershed_geom_nlcd = watershed_for_nlcd.geometry.iloc[0]

    # Perform the clip
    nlcd_clipped, nlcd_clip_transform = mask(
        src,
        [watershed_geom_nlcd],
        crop=True,
        nodata=0
    )

    # Update metadata
    nlcd_clip_meta = src.meta.copy()
    nlcd_clip_meta.update({
        'height': nlcd_clipped.shape[1],
        'width': nlcd_clipped.shape[2],
        'transform': nlcd_clip_transform,
        'nodata': 0
    })

# Extract the data array
nlcd_clipped_data = nlcd_clipped[0]

print(f"NLCD clipped to watershed:")
print(f"  Original size: {nlcd_data.shape}")
print(f"  Clipped size: {nlcd_clipped_data.shape}")
print(f"  Valid pixels: {(nlcd_clipped_data > 0).sum():,}")

NLCD clipped to watershed:
  Original size: (354, 535)
  Clipped size: (231, 414)
  Valid pixels: 54,044


Same clipping approach as the DEM, but notice one extra step: we reproject the watershed to match the NLCD's CRS before clipping. The NLCD might be in a different projection than the DEM, so we can't assume they share the same coordinate system.

## Part 8: Elevation Analysis 📊

### Calculate Elevation Statistics

In [ ]:
# Calculate elevation statistics
valid_elevations = dem_clipped_masked[~dem_clipped_masked.mask]

# Conversion factor used for the rest of Part 8 and the dashboard
M_TO_FT = 3.281

print("ELEVATION STATISTICS FOR WATERSHED")
print("="*50)
print(f"Minimum elevation: {valid_elevations.min():.1f} m  ({valid_elevations.min()*M_TO_FT:.1f} ft)")
print(f"Maximum elevation: {valid_elevations.max():.1f} m  ({valid_elevations.max()*M_TO_FT:.1f} ft)")
print(f"Mean elevation: {valid_elevations.mean():.1f} m  ({valid_elevations.mean()*M_TO_FT:.1f} ft)")
print(f"Median elevation: {np.median(valid_elevations):.1f} m  ({np.median(valid_elevations)*M_TO_FT:.1f} ft)")
print(f"Standard deviation: {valid_elevations.std():.1f} m  ({valid_elevations.std()*M_TO_FT:.1f} ft)")

elev_range = valid_elevations.max() - valid_elevations.min()
print(f"\nElevation range: {elev_range:.1f} m  ({elev_range*M_TO_FT:.1f} ft)")
print(f"Relief ratio: {elev_range/1000:.3f} m/km")

### Reading elevation statistics

These numbers characterize the terrain within your watershed:
- **Min/Max** tell you the elevation range (relief)
- **Mean** gives you the average elevation
- **Standard deviation** tells you how variable the terrain is - a flat watershed has low std dev, a mountainous one has high

The feet conversions (multiplying meters by 3.281) are included since most H&H work in the US uses feet.

If you want to add more statistics, ask your AI: *"How would I add the 10th and 90th percentile elevations to this summary?"*

### Create Elevation Bands

In [ ]:
# Define elevation bands for visualization and analysis
n_bands = 5
band_edges = np.linspace(valid_elevations.min(), valid_elevations.max(), n_bands + 1)

# Create elevation band raster
elevation_bands = np.zeros_like(dem_clipped_masked)
for i in range(n_bands):
    mask_band = (dem_clipped_masked >= band_edges[i]) & (dem_clipped_masked < band_edges[i+1])
    elevation_bands[mask_band] = i + 1

# Calculate area per band
print("\nELEVATION BANDS")
print("="*50)

# Pixel area in square meters, from the actual clipped DEM transform
pixel_area_m2 = abs(dem_clip_transform[0] * dem_clip_transform[4])

for i in range(n_bands):
    band_pixels = (elevation_bands == i + 1).sum()
    band_area_km2 = (band_pixels * pixel_area_m2) / 1_000_000
    band_percent = (band_pixels / (elevation_bands > 0).sum()) * 100

    lo_m, hi_m = band_edges[i], band_edges[i+1]
    print(f"Band {i+1}: {lo_m:.0f} - {hi_m:.0f} m  ({lo_m*M_TO_FT:.0f} - {hi_m*M_TO_FT:.0f} ft)")
    print(f"  Area: {band_area_km2:.2f} km² ({band_percent:.1f}%)")

### What are elevation bands for?

Elevation bands split the watershed into horizontal slices by elevation. This is useful for:
- **Snow hydrology** - estimating where rain vs. snow falls at different elevations
- **Temperature lapse rate** corrections
- **Vegetation zone** analysis
- **Visualizing terrain distribution** across the watershed

The code uses `np.linspace` to create evenly spaced band boundaries, then assigns each pixel to a band.

### Visualize Elevation

In [ ]:
# Create elevation visualization (panels stacked vertically)
fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(10, 14))

# Top: Continuous elevation, shown in feet
dem_clipped_ft = dem_clipped_masked * M_TO_FT
im1 = ax1.imshow(dem_clipped_ft, cmap='terrain', aspect='equal')
plt.colorbar(im1, ax=ax1, label='Elevation (ft)', shrink=0.8)
ax1.set_title('Digital Elevation Model', fontsize=14)
ax1.set_xticks([])
ax1.set_yticks([])

# Bottom: Elevation bands, with a colorbar labelled by elevation range in feet
colors_bands = ['#2E7D32', '#66BB6A', '#FDD835', '#FB8C00', '#D84315']  # Green to red
cmap_bands = ListedColormap(colors_bands[:n_bands])
im2 = ax2.imshow(np.ma.masked_where(elevation_bands == 0, elevation_bands),
                 cmap=cmap_bands, aspect='equal', vmin=0.5, vmax=n_bands + 0.5)

cbar2 = plt.colorbar(im2, ax=ax2, ticks=range(1, n_bands + 1), shrink=0.8,
                     label='Elevation band (ft)')
cbar2.set_ticklabels([f'{band_edges[i]*M_TO_FT:.0f}-{band_edges[i+1]*M_TO_FT:.0f}'
                      for i in range(n_bands)])

ax2.set_title('Elevation Bands', fontsize=14)
ax2.set_xticks([])
ax2.set_yticks([])

plt.suptitle(f'Watershed Elevation Analysis: HUC {target_huc}', fontsize=16)
plt.tight_layout()
plt.show()

The top panel shows continuous elevation in feet, and the bottom panel shows the same terrain grouped into elevation bands. If you want to change the number of bands, modify `n_bands = 5` in the code above and re-run. More bands give finer resolution but can make the map harder to read.

To change the band colors, edit the `colors_bands` list. You can ask: *"Give me a list of 7 hex color codes that go from green to red for an elevation map."*

## Part 9: Land Cover Analysis 🌿

### Calculate Land Cover Statistics

In [23]:
# Calculate land cover statistics
print("LAND COVER STATISTICS FOR WATERSHED")
print("="*60)

# Get unique values and counts
unique_lc, counts_lc = np.unique(nlcd_clipped_data[nlcd_clipped_data > 0], return_counts=True)

# Calculate areas (assuming 30m resolution for NLCD)
pixel_area_nlcd = 30 * 30  # square meters
total_pixels = counts_lc.sum()
total_area_km2 = (total_pixels * pixel_area_nlcd) / 1_000_000

# Create summary table
lc_stats = []

for value, count in zip(unique_lc, counts_lc):
    area_km2 = (count * pixel_area_nlcd) / 1_000_000
    percentage = (count / total_pixels) * 100

    lc_stats.append({
        'Code': value,
        'Class': nlcd_classes.get(value, 'Unknown'),
        'Pixels': count,
        'Area_km2': area_km2,
        'Percent': percentage
    })

# Create DataFrame and sort by area
lc_df = pd.DataFrame(lc_stats)
lc_df = lc_df.sort_values('Area_km2', ascending=False)

print(f"Total watershed area: {total_area_km2:.2f} km²\n")
print(lc_df.to_string(index=False))

LAND COVER STATISTICS FOR WATERSHED
Total watershed area: 48.64 km²

 Code                        Class  Pixels  Area_km2   Percent
   71         Grassland/Herbaceous   35912   32.3208 66.449560
   22     Developed, Low Intensity    9372    8.4348 17.341426
   23  Developed, Medium Intensity    4109    3.6981  7.603064
   21        Developed, Open Space    3042    2.7378  5.628747
   31                  Barren Land     533    0.4797  0.986233
   24    Developed, High Intensity     448    0.4032  0.828954
   95 Emergent Herbaceous Wetlands     430    0.3870  0.795648
   82             Cultivated Crops     110    0.0990  0.203538
   81                  Pasture/Hay      72    0.0648  0.133225
   11                   Open Water      14    0.0126  0.025905
   52                  Shrub/Scrub       2    0.0018  0.003701


### Reading the land cover table

The table shows each land cover type in the watershed, sorted by area. The key columns:
- **Pixels** - how many 30m x 30m pixels have this land cover
- **Area_km2** - the area in square kilometers
- **Percent** - what fraction of the watershed this class covers

This breakdown is exactly what you'd need for a watershed characterization in an H&H report. The dominant land cover class drives many modeling decisions (curve numbers, roughness, evapotranspiration).

### Visualize Land Cover

In [ ]:
# Create land cover map
fig, ax1 = plt.subplots(figsize=(11, 9))

# Build a colormap from the NLCD classes that actually appear in the watershed
unique_in_watershed = nlcd_clipped_data[nlcd_clipped_data > 0].flatten()
unique_classes = np.unique(unique_in_watershed)

nlcd_display = np.zeros_like(nlcd_clipped_data)
color_list = []
labels_list = []

for i, class_val in enumerate(unique_classes):
    nlcd_display[nlcd_clipped_data == class_val] = i + 1
    color_list.append(nlcd_colors.get(class_val, '#808080'))
    labels_list.append(f"{class_val}: {nlcd_classes.get(class_val, 'Unknown')}")

cmap_nlcd = ListedColormap(color_list)
im = ax1.imshow(np.ma.masked_where(nlcd_display == 0, nlcd_display),
                cmap=cmap_nlcd, vmin=1, vmax=len(unique_classes))

# Legend in the lower-right corner
legend_elements = [mpatches.Patch(facecolor=color_list[i], label=labels_list[i])
                   for i in range(len(unique_classes))]
ax1.legend(handles=legend_elements, loc='lower right', fontsize=8)

ax1.set_title(f'Land Cover Classification: HUC {target_huc}', fontsize=14)
ax1.set_xticks([])
ax1.set_yticks([])

plt.tight_layout()
plt.show()

The map uses the standard NLCD color scheme, so it looks familiar if you've worked with this data in GIS. The legend in the lower-right corner lists every land cover class present in the watershed.

If the legend covers part of the map, you can ask your AI: *"How do I move a matplotlib legend outside the plot area?"*

## Part 10: Slope Analysis 🏔️

Calculate slope from DEM for runoff analysis:

In [ ]:
# Calculate slope from the clipped DEM
#
# The DEM is in a geographic CRS (degrees), so the pixel size from the
# transform is in degrees, not meters. We need to convert to meters for
# np.gradient to produce correct rise-over-run values.

import rasterio.transform

# Get pixel dimensions in degrees from the clipped DEM transform
dx_degrees = dem_clip_transform.a
dy_degrees = abs(dem_clip_transform.e)

# Find the center latitude of the clipped raster to convert degrees -> meters
height = dem_clip_meta["height"]
width = dem_clip_meta["width"]
left, bottom, right, top = rasterio.transform.array_bounds(height, width, dem_clip_transform)
mean_lat_rad = np.deg2rad((top + bottom) / 2)

# Approximate conversion at this latitude
# 1 degree latitude  ~ 111,139 m everywhere
# 1 degree longitude ~ 111,319 m * cos(latitude)
dx_m = dx_degrees * 111319.0 * np.cos(mean_lat_rad)
dy_m = dy_degrees * 111139.0

print(f"Pixel size for slope calculation: {dx_m:.2f} m (E-W) x {dy_m:.2f} m (N-S)")

# np.gradient cannot handle masked arrays correctly -- it computes huge
# gradients at the boundary between real elevations and NoData fill values.
# Fill masked pixels with the local mean first, then re-mask afterward.
dem_filled = dem_clipped_masked.filled(fill_value=np.nanmean(dem_clipped_masked))

# Compute gradients (np.gradient returns row-first, so dz_dy comes first)
dz_dy, dz_dx = np.gradient(dem_filled, dy_m, dx_m)

# Slope as a percent grade: rise over run, times 100
slope_percent = np.sqrt(dz_dx**2 + dz_dy**2) * 100

# Re-apply the original mask
slope_percent = np.ma.masked_array(slope_percent, mask=dem_clipped_masked.mask)
valid_slopes = slope_percent.compressed()

print("\nSLOPE STATISTICS")
print("="*40)
print(f"Minimum slope: {valid_slopes.min():.1f}%")
print(f"Maximum slope: {valid_slopes.max():.1f}%")
print(f"Mean slope: {valid_slopes.mean():.1f}%")
print(f"Median slope: {np.median(valid_slopes):.1f}%")

# Classify slopes for H&H analysis (percent grade)
print("\nSLOPE CLASSES FOR RUNOFF")
print("="*40)
flat = (valid_slopes < 2).sum() / len(valid_slopes) * 100
gentle = ((valid_slopes >= 2) & (valid_slopes < 5)).sum() / len(valid_slopes) * 100
moderate = ((valid_slopes >= 5) & (valid_slopes < 10)).sum() / len(valid_slopes) * 100
steep = ((valid_slopes >= 10) & (valid_slopes < 20)).sum() / len(valid_slopes) * 100
very_steep = (valid_slopes >= 20).sum() / len(valid_slopes) * 100

print(f"Flat (< 2%): {flat:.1f}% of area")
print(f"Gentle (2-5%): {gentle:.1f}% of area")
print(f"Moderate (5-10%): {moderate:.1f}% of area")
print(f"Steep (10-20%): {steep:.1f}% of area")
print(f"Very steep (> 20%): {very_steep:.1f}% of area")

### How slope is calculated

The code uses `np.gradient` to estimate the rate of elevation change in both the x (east-west) and y (north-south) directions. Slope is then reported as a percent grade: rise over run, multiplied by 100. A 5% slope means the ground rises 5 feet for every 100 feet of horizontal distance. Percent grade is the convention used in most drainage and H&H design work.

Since this DEM is in geographic coordinates (degrees), the pixel sizes are first converted from degrees to approximate meters using the latitude of the watershed. Without this conversion, the slopes would be wildly incorrect.

The slope classes (flat, gentle, moderate, steep, very steep) are commonly used thresholds in H&H work for erosion risk assessment and runoff estimation.

### Visualize Slope

In [ ]:
# Visualize slope
fig, ax = plt.subplots(figsize=(10, 8))

im = ax.imshow(slope_percent, cmap='YlOrRd', aspect='equal', vmin=0, vmax=30)
plt.colorbar(im, ax=ax, label='Slope (%)', shrink=0.8)

ax.set_title('Terrain Slope Analysis', fontsize=14)
ax.set_xticks([])
ax.set_yticks([])

# Add text with key statistics
textstr = f'Mean: {valid_slopes.mean():.1f}%\n' \
          f'Max: {valid_slopes.max():.1f}%\n' \
          f'Steep areas (>10%): {(steep + very_steep):.1f}%'
props = dict(boxstyle='round', facecolor='wheat', alpha=0.8)
ax.text(0.02, 0.98, textstr, transform=ax.transAxes, fontsize=10,
        verticalalignment='top', bbox=props)

plt.tight_layout()
plt.show()

The `YlOrRd` (yellow-orange-red) colormap makes steep areas jump out visually. If you prefer a different look, try `cmap='Reds'` or `cmap='RdYlGn_r'` (reversed red-yellow-green).

You can also change the `vmax=30` parameter to adjust the color scale. If your watershed has slopes above 30%, increase it so the full range is visible.

## Part 11: H&H Applications - Curve Numbers 💧

Assign SCS Curve Numbers based on land cover:

In [27]:
# SCS Curve Numbers for Hydrologic Soil Group B (adjust as needed)
# For simplicity we are assuming the entire watershed can be represented
# by HSG B (that is usually not the case in real watersheds)
cn_lookup = {
    11: 100,  # Open Water
    12: 100,  # Ice/Snow
    21: 61,   # Developed, Open Space
    22: 75,   # Developed, Low Intensity
    23: 83,   # Developed, Medium Intensity
    24: 93,   # Developed, High Intensity
    31: 77,   # Barren Land
    41: 55,   # Deciduous Forest
    42: 55,   # Evergreen Forest
    43: 55,   # Mixed Forest
    52: 48,   # Shrub/Scrub
    71: 65,   # Grassland
    81: 69,   # Pasture/Hay
    82: 75,   # Cultivated Crops
    90: 85,   # Woody Wetlands
    95: 85    # Herbaceous Wetlands
}

# Calculate weighted curve number
total_cn_area = 0
total_area = 0

print("CURVE NUMBER ANALYSIS")
print("="*60)
print(f"{'Land Cover':<30} {'Area (km²)':<10} {'CN':<5} {'Weighted':<10}")
print("-"*60)

for _, row in lc_df.iterrows():
    cn = cn_lookup.get(row['Code'], 70)  # Default CN of 70
    weighted = row['Area_km2'] * cn
    total_cn_area += weighted
    total_area += row['Area_km2']

    print(f"{row['Class']:<30} {row['Area_km2']:<10.2f} {cn:<5} {weighted:<10.1f}")

# Calculate composite CN
composite_cn = total_cn_area / total_area if total_area > 0 else 0

print("-"*60)
print(f"{'COMPOSITE CURVE NUMBER:':<30} {'':<10} {composite_cn:<5.1f}")
print("\n💡 This CN can be used for runoff calculations in HEC-HMS or similar models")

CURVE NUMBER ANALYSIS
Land Cover                     Area (km²) CN    Weighted  
------------------------------------------------------------
Grassland/Herbaceous           32.32      65    2100.9    
Developed, Low Intensity       8.43       75    632.6     
Developed, Medium Intensity    3.70       83    306.9     
Developed, Open Space          2.74       61    167.0     
Barren Land                    0.48       77    36.9      
Developed, High Intensity      0.40       93    37.5      
Emergent Herbaceous Wetlands   0.39       85    32.9      
Cultivated Crops               0.10       75    7.4       
Pasture/Hay                    0.06       69    4.5       
Open Water                     0.01       100   1.3       
Shrub/Scrub                    0.00       48    0.1       
------------------------------------------------------------
COMPOSITE CURVE NUMBER:                   68.4 

💡 This CN can be used for runoff calculations in HEC-HMS or similar models


### What is the composite Curve Number?

The SCS Curve Number method estimates runoff based on land cover and soil type. Each land cover class gets a CN value (higher = more runoff), and the composite CN is the area-weighted average across the entire watershed.

This composite CN can be plugged directly into HEC-HMS or used in hand calculations with TR-55 methods. A CN of ~70 is moderate; below 60 means the watershed absorbs a lot of rainfall, above 80 means most rainfall becomes runoff.

Keep in mind that the CN values here assume Hydrologic Soil Group B throughout the watershed. In a real project, you'd want to overlay a soils map (like SSURGO) to assign the correct soil group to each pixel.

## Part 12: Professional Integrated Map 🎨

Create a comprehensive map combining all analyses:

In [ ]:
# Create professional integrated visualization
fig = plt.figure(figsize=(18, 12))

# Create grid for subplots
gs = fig.add_gridspec(3, 3, hspace=0.3, wspace=0.3)

# 1. DEM (feet)
ax1 = fig.add_subplot(gs[0, 0])
im1 = ax1.imshow(dem_clipped_masked * M_TO_FT, cmap='terrain', aspect='equal')
plt.colorbar(im1, ax=ax1, label='Elev (ft)', shrink=0.8)
ax1.set_title('Digital Elevation Model', fontsize=10)
ax1.axis('off')

# 2. Slope (percent grade)
ax2 = fig.add_subplot(gs[0, 1])
im2 = ax2.imshow(slope_percent, cmap='YlOrRd', aspect='equal', vmin=0, vmax=30)
plt.colorbar(im2, ax=ax2, label='Slope (%)', shrink=0.8)
ax2.set_title('Terrain Slope', fontsize=10)
ax2.axis('off')

# 3. Land Cover
ax3 = fig.add_subplot(gs[0, 2])
im3 = ax3.imshow(np.ma.masked_where(nlcd_display == 0, nlcd_display),
                 cmap=cmap_nlcd, vmin=1, vmax=len(unique_classes))
ax3.set_title('Land Cover', fontsize=10)
ax3.axis('off')

# 4. Elevation Histogram (feet)
ax4 = fig.add_subplot(gs[1, 0])
ax4.hist(valid_elevations * M_TO_FT, bins=30, color='skyblue', edgecolor='darkblue', alpha=0.7)
ax4.set_xlabel('Elevation (ft)', fontsize=9)
ax4.set_ylabel('Frequency', fontsize=9)
ax4.set_title('Elevation Distribution', fontsize=10)
ax4.grid(True, alpha=0.3)

# 5. Slope Classes Bar Chart
ax5 = fig.add_subplot(gs[1, 1])
slope_classes = ['Flat\n(<2%)', 'Gentle\n(2-5%)', 'Moderate\n(5-10%)',
                 'Steep\n(10-20%)', 'Very Steep\n(>20%)']
slope_percentages = [flat, gentle, moderate, steep, very_steep]
bars = ax5.bar(slope_classes, slope_percentages,
               color=['#2E7D32', '#66BB6A', '#FDD835', '#FB8C00', '#D84315'])
ax5.set_ylabel('Area (%)', fontsize=9)
ax5.set_title('Slope Classification', fontsize=10)
ax5.grid(True, alpha=0.3, axis='y')
for bar, pct in zip(bars, slope_percentages):
    ax5.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.5,
             f'{pct:.1f}%', ha='center', fontsize=8)

# 6. Top Land Covers (horizontal bar)
ax6 = fig.add_subplot(gs[1, 2])
top_5 = lc_df.head(5).iloc[::-1]  # reverse so the largest sits at the top
colors_bar = [nlcd_colors.get(code, '#808080') for code in top_5['Code']]
ax6.barh(top_5['Class'], top_5['Percent'], color=colors_bar, edgecolor='gray')
ax6.set_xlabel('Area (%)', fontsize=9)
ax6.set_title('Top 5 Land Covers', fontsize=10)
ax6.tick_params(labelsize=8)
ax6.grid(True, alpha=0.3, axis='x')

# 7. Summary Statistics Table
ax7 = fig.add_subplot(gs[2, :])
ax7.axis('tight')
ax7.axis('off')

summary_data = [
    ['Parameter', 'Value', 'Units'],
    ['Watershed Area', f'{total_area_km2:.2f}', 'km²'],
    ['Elevation Range', f'{valid_elevations.min()*M_TO_FT:.0f} - {valid_elevations.max()*M_TO_FT:.0f}', 'ft'],
    ['Mean Elevation', f'{valid_elevations.mean()*M_TO_FT:.1f}', 'ft'],
    ['Mean Slope', f'{valid_slopes.mean():.1f}', '%'],
    ['Composite CN', f'{composite_cn:.1f}', '-'],
    ['Dominant Land Cover', lc_df.iloc[0]['Class'], f"{lc_df.iloc[0]['Percent']:.1f}%"]
]

table = ax7.table(cellText=summary_data[1:], colLabels=summary_data[0],
                  cellLoc='center', loc='center',
                  colWidths=[0.3, 0.3, 0.2])
table.auto_set_font_size(False)
table.set_fontsize(9)
table.scale(1, 1.5)

# Style header row
for i in range(3):
    table[(0, i)].set_facecolor('#4CAF50')
    table[(0, i)].set_text_props(weight='bold', color='white')

# Main title
fig.suptitle(f'Watershed Raster Analysis Dashboard: HUC {target_huc}', fontsize=16, y=0.98)

# Subtitle with data sources
fig.text(0.5, 0.95, 'Data Sources: USGS DEM, NLCD 2024 | Analysis: 2024',
         ha='center', fontsize=10, style='italic')

plt.tight_layout()
plt.show()

print("Professional integrated analysis map created!")

### The dashboard at a glance

This six-panel figure summarizes everything we've calculated for the watershed:
- **Top row**: DEM, slope map, and land cover classification (the spatial patterns)
- **Middle row**: elevation histogram, slope class bar chart, and top land covers pie chart (the distributions)
- **Bottom**: summary statistics table with the key numbers you'd include in a report

This kind of dashboard is a great deliverable for early-stage project work. You can customize the layout, colors, and content by modifying the subplot grid and the data shown in each panel.

Want to save it as a high-resolution image? Add `plt.savefig('watershed_dashboard.png', dpi=300, bbox_inches='tight')` before `plt.show()`.

## Practice Exercises 🎯

### Exercise 1: Multi-Watershed Raster Analysis
Analyze elevation and land cover for 3 different watersheds.

In [ ]:
# EXERCISE 1: Compare raster statistics across watersheds
# Your code here:

# Step 1: Select 3 watersheds

# Step 2: For each, clip DEM and NLCD

# Step 3: Calculate statistics

# Step 4: Create comparison table


### Exercise 2: Aspect Analysis
Calculate and visualize terrain aspect (direction of slope).

In [ ]:
# EXERCISE 2: Calculate aspect from DEM
# Your code here:

# Hint: Use np.arctan2(dz_dy, dz_dx)
# Convert to degrees and classify (N, NE, E, SE, S, SW, W, NW)


### Exercise 3: Elevation-Land Cover Relationship
Analyze how land cover varies with elevation.

In [ ]:
# EXERCISE 3: Land cover by elevation band
# Your code here:

# Step 1: Create elevation bands

# Step 2: For each band, extract land cover

# Step 3: Calculate dominant land cover per band

# Step 4: Visualize results


### Exercise 4: Runoff Potential Map
Create a map combining slope and land cover to show runoff potential.

In [ ]:
# EXERCISE 4: Runoff potential analysis
# Your code here:

# Combine high slopes + high CN areas
# Create risk categories (Low, Medium, High)
# Visualize as a new raster


## That's Module 4 Done!

You now have a complete spatial analysis toolkit. Between Module 3 (vector) and Module 4 (raster), you can:

- Define WHERE your watershed is (vector boundaries)
- Analyze WHAT'S INSIDE it (raster surfaces)
- Understand HOW water moves (terrain analysis)
- Predict runoff response (land cover and curve numbers)

### What you can do now:

**Raster Data** - Load DEMs and land cover data, understand resolution and NoData, clip rasters to areas of interest

**Terrain Analysis** - Calculate elevation statistics, generate slope maps, create elevation bands, identify steep areas

**Land Cover** - Classify and quantify land uses, calculate composite curve numbers, link land cover to runoff potential

**Integrated Analysis** - Combine vector watersheds with raster data, create professional dashboard maps, generate model inputs

### Key things to remember:

1. **Rasters are grids of values** - each pixel holds a measurement
2. **Resolution matters** - balance detail with file size
3. **CRS alignment is critical** - always match coordinate systems before clipping
4. **Combine analyses** - terrain + land cover = runoff understanding

### Quick reference:

```python
rasterio.open()          # Load raster
src.read()              # Read data array
mask()                  # Clip to boundary
np.gradient()           # Calculate slope
zonal_stats()           # Statistics by zone
plt.imshow()            # Visualize
```

### What's next

In Module 5, you'll learn to pull real USGS gauge data programmatically - connecting your spatial analysis with actual measurements for model calibration.